### ISSUE TITLE
Allow admins to change the date format

### ISSUE THREAD

**Issue by kjantin** [huridocs/uwazi#979]
_"As an administrator, I want to change the format of the date field display from **May 8, 2017** to **8 May 2017**."_

Related to #815

**Comment by kjantin** [comment]
The date format is being added according to the language of the instance. We will look into more nuanced date formatting for the English language (EN-British, EN-American, etc).

**Comment by danicatalan** [comment]
## tl;dr

Three different issues:

**Input style** 

- [ ] Allow admin to choose 3 different styles to write dates on inputs:
`YYYY-MM-DD`
`DD-MM-YYYY`
`MM-DD-YYYY`

- [ ] Allow admin to choose 3 different styles of separators:
`/`
`-`
`.`

**Date styles in English**

- [ ] Set `English (UK)` as default format (_24 May 2017_).
- [ ] If `MM-DD-YYYY`, use `English (US)` (_May 24, 2017_).

**Human-readable format switcher**

- [ ] Allow admins switch between render dates in numbers (_2017/05/24_) or in human-readable way (_24 May 2017_) on cards and sidepanel.

- - - - - 

## Context

First of all, we have two contexts where data is rendered:

Read dates: Users can find dates throughout the app, mostly on cards and sidepanels. When reading dates, we used a library called Moment.js to translate plain dates (*2017-05-24T10:42:58+02:00*) into more semantic ones (*May 24, 2017 10:42 AM*). Also we support language format, so Moment.js will translate plain dates into other languages than English (*24 de mayo de 2017 10:42*). This is great because now we write dates in English and they are rendered in the better language + format for the user.

Write dates: Users want to introduce dates on a friendly format (#1024). Formats like *24/05/2017* are much more common and easier for storing data than *May 24, 2017*. At this moment we keep the same input > output on English, but in other languages the input is in English and the output is in Spanish. This suggestion can standarize all the languages to the same common input format.

![date-format](https://cloud.githubusercontent.com/assets/6364153/26554743/1c192180-4492-11e7-9d18-8766734cbc27.png)

## Format

We have a lot of formats available on Moment.js. Most of them are dates with hours and seconds or related time ("Last Monday at 10:42"). In our case we have two interesting formats:

    moment().format('L');    // 05/24/2017
    moment().format('LL');   // May 24, 2017

While numbers are great for writing dates, sometimes is preferable to use a more human-readable format to display dates. Anyway, I think both formats can work pretty well for reading dates, so we can let admin decide which one they want to use.

After the decision about which kind of date we use for every context, we need to choose the location of the language. Sometimes is easy because one language only supports one format. But as you can see in the next image, [there are languages with more than one way to represent dates](https://cloud.githubusercontent.com/assets/6364153/26554762/2c8be142-4492-11e7-8c08-61decb9ff352.png).

For example, English US is **May 24, 2017** and English UK is **24 May 2017**. I've been researching about this and affects at differents telements while rendering the dates:

- For L formats (numbers), it may changes the components order and the separator of those components.
- For LL formats (words), it changes the way the components are named and their order.

We can solve this problem with sub-language locations (but we will duplicated the content with mayor benefits) or try to solve this with a smart configuration system.

First of all, the admin select the input format. Input will be always on L format (numbers) and we will allow three different orders styles:

- YYYY-MM-DD: 2017/05/24. This is the default one suggested by ISO 8601. Mostly used in Asia but also it's getting official in more and more countries and systems because of the ISO standard.
- DD-MM-YYYY: 24/05/2017. Most common in the world. Europe, Africa, Latin America, South Asia and Oceania uses this. A lot of countries have this style official and also supports YMD because of the ISO standard.
- MM-DD-YYYY: 05/24/2017. USA. Why...

We also need support for separators. In most countries with YY-MM-DDDD format they use dashes instead of slashes.

Anyway, the input selection will affect to the human-readable date.

- MM-DD-YYYY is selected, English US overrides other English.
- Any other format will fallback to their language style.
- We will need to check Arab when supported, because they have different styles between countries too.

**Comment by kjantin** [comment]
Looks great!

**Comment by txau** [comment]
I'd like @RafaPolit input on this topic.

**Comment by RafaPolit** [comment]
@txau @danicatalan Personally I would not over-complicate things at this point.

I think the real problem is date INPUT.

So my take would be to implement the ability for the admin to configure globally the input mask chosen between:
YYYY-MM-DD
DD-MM-YYYY
MM-DD-YYYY

And provide two options for separators: / and -

I don't think the period should be included, the more options the more time it will take to get this out working.  Does anyone use period as date separator like 2017.05.06? I have never seen that.

Other than that, the only thing I would do is unify so that there is no inconsistencies on the rendered dates and combine them all under MomentJS governed by the language selected.

I certainly think we:
- SHOULD NOT do Single digit support: if the mask is clearly stated MM, put the 0 in front.  Current problem is LACK OF INFO as to how to do it plus the problem that you are inputting letters as well.  But if the mask has two digits, lets not complicate the things.
- SHOULD NOT do  Admin selectable readable format:  the readable format is good as it is, even for UK vs US.  As long as there is text, everything is clear, at least from my point of view.  So "24 of may, 2017" and "May 24, 2017" are really perfectly understandable.  The issue arises when you display only numbers.  But if we keep the LL format, I believe we shouldn't change that.
- SHOULD NOT do sub-language localization because we are going to display the month in letters, so the need for this is reduced to almost unnecessary.
- SHOULD NOT derive rendered format from the input method selected (if MM-DD then US, if DD-MM then UK), rather from the language selected by the user, as it is on the cards right now.
- SHOULD NOT do anything anywhere, except for the input field.

So, to sum things up:
- Render every date using MomentJS LL rendering ("May 24, 2017" for EN, or "24 de may, 2017" for ES, etc.) without any other mask available, not configurable at all.  This is almost what already is implemented except for the date input when showing its value. 
- Render the date input always with the admin selected mask, never in letters (to avoid changing the mask when you are inputting data vs when you are just reading the input field).
- Put a placeholder or even the format into the label of the date field to prevent any confusion: Date label (MM-DD-YYYY)
- Make date settings configurable only with 3 date formats and 2 separators.

I really think we are overcomplicating this issue which is a date-input-only problem.

That is my take.  I honestly thought we had already discussed this in the sprint planning, but it appears everyone just thought that their way was the best approach.

Maybe later on we can keep on evolving this, but this will have users inputting numbers fairly soon.  The 'full fledged' implementation above would involve a lot of work.

**Comment by danicatalan** [comment]
Just going to implement #1053 as @RafaPolit suggest.

Closing this.
